Revise SDL-0282 Screen Manager Alert Manager (#1110)
smartdevicelink/sdl_evolution: proposals/0282-screen-manager-alerts.md

# Screen Manager Alert Manager

* Proposal: [SDL-0282](0282-screen-manager-alerts.md)

> * Author: [Joel Fischer (https://github.com/joeljfischer), Nicole Yarroch (https://github.com/NicoleYarroch), Julian Kast (https://github.com/JulianKast)]

Reviewer: ```suggestion
* Author: [Joel Fischer](https://github.com/joeljfischer), [Nicole Yarroch](https://github.com/NicoleYarroch), [Julian Kast](https://github.com/JulianKast)
```

 @param spokenString The string to be spoken by the system speech synthesizer
 @param phoneticType Must be one of `SAPI_PHONEMES`, `LHPLUS_PHONEMES`, `TEXT`, or `PRE_RECORDED` or no object will be created
 */
- (instancetype)initWithPhoneticSpeechSynthesizerString:(NSString *)phoneticString phoneticType:(SDLSpeechCapabilities)phoneticType;

/**

>  Add additional SDLFiles holding data or pointing to a file on the file system. When this object is passed to an `Alert` or `Speak`, the file will be uploaded if it is not already, then played if the system supports that feature.

Reviewer: ```suggestion
 Add additional `SDLFile`s holding data or pointing to a file on the file system. When this object is passed to an `Alert` or `Speak`, the file will be uploaded if it is not already, then played if the system supports that feature.
```

        Builder setDefaultTimeout(int defaultTimeout) { /* Imp */ }
        Builder setTimeout(int timeout) { /* Imp */ }
        Builder setSoftButtons(List<SoftButtonObject> softButtons) { /* Imp */ }
        Builder setAudio(AlertAudioData audio) { /* Imp */ }
        Builder setIcon(SdlArtwork icon) { /* Imp */ }
    }

    // All vars have getters and setters to match iOS read / write

>     public void cancel() {}

Reviewer 2: ```suggestion
    public void cancel() {}
    
    // Creates a deep copy of the object
    @Override
    public AlertView clone() 
```

##### Java
```java
public class BaseScreenManager {

>     // Everything already there

Reviewer 2: ```suggestion
```

    public void presentAlert(AlertView alert, AlertCompletionListener listener)
}

public interface AlertCompletionListener {
    /**
     * Returns whether an Alert operation was successful or not along with tryAgainTime
     * @param success - Boolean that is True if Operation was a success, False otherwise.
     * @param tryAgainTime - Amount of time (in seconds) that an app must wait before resending an alert.
     */

>     void onComplete(boolean success, Integer tryAgainTime);

Reviewer 2: Should we separate this into two blocks of code, I don't want to give the impression that the AlertCompletionListener is in the same file as BaseScreenManager, as it is its own file

Author: Separated into two blocks

- The internal alert manager will observe the `AlertResponse` to know when the alert has finished presenting, and then call the `completionHandler`.
- The internal alert manager will always send the alert, even if the system context is not MAIN. If the `AlertResponse` returns a failure to present, it will call the `completionHandler` with the error.
- The developer will not be notified when the alert appears on the screen, assuming no error occurred – see alternative #1 for possible ways to do that.

> - The `SDLAlertManager` sub-manager will use queues to manage alert related requests, similar to the implementation in the `SDLChoiceSetManager`. The queue is serial and if an alert is sent while another alert is currently presented, the newest alert will not be sent until the module dismisses the previous alert.

Reviewer: ```suggestion
- The `SDLAlertManager` sub-manager will use queues to manage alert related requests, similar to the implementation in the `SDLChoiceSetManager`. The queue is serial and if an alert is sent while another alert is currently presented, the newest alert will not be sent until the module dismisses the previous alert or `cancel` is called on the alert view by the developer on RPC 6.0+ systems.
```
